### ISSUE TITLE
Write transformed kernels to file

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#201]
This is a follow-on to #185 which implements a first (OpenACC) kernel transformation.
In that issue we modify the kernel AST. In this issue we will add support for writing transformed kernels to disk.

**Comment by arporter** [comment]
We will need to add/extend the compilation testing to kernels that have been modified.

**Comment by arporter** [comment]
Since we have to know the kernel (module) names at PSy-code generation time, I'm now passing the options that control kernel naming into the PSy object constructor. My plan is to generate the kernel filenames (and the kernel code itself) when `psy.gen` is called.

**Comment by arporter** [comment]
I still need to ensure that the kernel and module are re-named in the generated PSy-layer code. I need to test this when the same kernel is used more than once in an Algorithm but only one instance is transformed.

**Comment by arporter** [comment]
(The PR that changes the management of the Config object is #215.)

**Comment by arporter** [comment]
I've realised I need to re-name any transformed kernel (to ensure we don't get a clash with an un-transformed version of that kernel that might be used elsewhere). I also need to take more trouble to ensure that the new names are unique. One possibility is to generate some sort of hash based on the Algorithm, Invoke and Kernel names as that combination must be unique.

**Comment by rupertford** [comment]
I agree that a hash of algorithm, invoke (position or name) and kernel is the way to go. This, of course assumes that we want unique for all kernels. It may be that we want the same optimisation for a kernel that is used multiple times in a 1) invoke, 2) algorithm, 3) application 4) some combination. Of course by treating each kernel individually we have full control, it just means that the same kernel is compiled multiple times, which may be an issue in large codes (but that is speculation).

**Comment by arporter** [comment]
The combination of module and subroutine names will uniquely identify a kernel in a given application. I think we had previously agreed that we would support two cases:
1. Making sure each transformed kernel is given a unique name;
2. Only transforming a given kernel once.

Since PSyclone only knows about the current Algorithm, the only way of ensuring that all of the kernels in an Application are given unique names (when transformed) is to store the state somewhere - i.e. in a given filesystem directory.

**Comment by TeranIvy** [comment]
Does renaming the kernel mean that the related `use` statement in the Algorithm will have to be re-written as well? 
I agree that module and subroutine name is usually enough for a given Algorithm if the target kernel is only called once. If it is called multiple times, however, then this would imply that all calls to that kernel need to use the transformed kernel instead of the original. If you want to keep the ability to "mix-and-match", then the invoke name is required as well (plus the new "use" statement in the Algorithm).
If there is more than one call to the kernel in the same invoke, the position in the schedule is likely to be required as well. 
Psyke forced me to think about this :-)

**Comment by rupertford** [comment]
Hi @TeranIvy. It would be the name in the PSy-layer that would need to change.

**Comment by arporter** [comment]
Yes, I have that bit working :-) It's just what name to choose that I'm working on...

**Comment by arporter** [comment]
For the case where we wish to ensure that a given kernel is only ever transformed once, the name of the new kernel must be uniquely defined by the name of the original - i.e. a hash of the module and subroutine names. As @TeranIvy has just noted (thanks!), if we wish to uniquely name a given kernel call then we will need the Algorithm module, the invoke name/index and the position of the kernel within the Schedule of that invoke. I can easily generate e.g. an MD5 hash of the string obtained by concatenating those various elements. It's not going to look very nice but then this is generated code (and I don't want to simply concatenate names because we'll rapidly crash into name-length limits in Fortran compilers).

**Comment by arporter** [comment]
After discussing with Rupert and giving it a lot of thought, I've gone for generating a random number (in the range [0,1000000) ) and converting that to hex. This creates a far friendlier looking name than adding in an MD5 hash e.g.: `testkern_mod` might become something like ` testkerne19bc_mod`.

**Comment by arporter** [comment]
Following yet more discussion, we've agreed that the risk that we generate two files with the same name is unacceptable. Therefore, we will use the filesystem to hold state between PSyclone runs and require that a single Application use a single kernel output directory. We can then check that directory for name clashes (and do clobber/no-clobber too). The `os.open` method can be made to obtain an exclusive lock on the target file and is atomic. However, there may be issues with this on parallel filesystems.

**Comment by arporter** [comment]
OK, we're back where we started (i.e. renaming kernels depending on the contents of a directory), albeit with an implementation that hopefully will work for parallel builds. I've just written the documentation to say that we decide a kernel has already been transformed based purely on its filename. Obviously that's not very robust. @rupertford should I compute an MD5 checksum or something and abort if the user has selected 'single' transformed kernel output and we've got differing versions of the same kernel?

**Comment by rupertford** [comment]
That makes sense to me. I don't like the idea of keeping a randomly optimised kernel so I agree that we should abort. Obviously we want the cheapest way of checking for a difference, whatever that is. I guess some form of caching will be needed at some point (in a separate issue).

**Comment by arporter** [comment]
Given that we already have the code for the transformed kernel in memory, I've decided the easiest way to compare with an existing kernel is to read that into memory too and then do a string comparison.
We can always review this in future if it becomes a bottleneck and perhaps start caching MD5 checksums or something.
I'm now just waiting on #215 to bring in changes to the management of the configuration object (because the way it's currently done is causing test failures for this branch).

**Comment by rupertford** [comment]
Merged PR #224 to master. Closing this issue.
